### ISSUE TITLE
Deep Neural Network Filter

### ISSUE THREAD

**Issue by yaitskov** [darktable-org/darktable#12316]
User selects pixels, which should belong to the filter and which should not.
DT builds DNN to recognize selected pixels taking into account their neighbors.

After training the DNN is applied to select pixels.

Filter parameters related to DNN:
- list of layers with number of neurons in each
- seed for random number generator
- optimization algorithm (adam, sgd, etc)
- checkbox whether to take into account pixel coordinates besides color channel values

**Comment by MStraeten** [comment]
Which Problem do you want to solve with that?

**Comment by yaitskov** [comment]
I am looking for a smooth way to select pixels on a complex photo, where same looking pixels are spread all over the photo, but belong to different objects of complicated form which are not separable by sane number of drawn curves.

e.g. black hair dyed in blond over white body skin and black hair roots and black top/jeans are mixed with black rock background.

**Comment by MStraeten** [comment]
do you have an idea how to train such DNN? 
Each module in the dt pixel pipe has different input data from raw data without rgb information over an unlimited to a limited colorspace. So you need to train for each of those intermediate data even those you cant display on a common display ;) 
keep in mind: in darktable you never see the output od a module - just the output of the whole pixelpipe

**Comment by jenshannoschwalm** [comment]
I also think there is room for having some sort of "content aware masking" in dt. First problems

1. Content would most likely involve whole image data and thus interfere with roi concept
2. Need for an interface to use those "content masks"

**Comment by spaceChRiS** [comment]
Probably related: #12295.

**Comment by jenshannoschwalm** [comment]
Yes, several issues basically on one topic

**Comment by yaitskov** [comment]
> do you have an idea how to train such DNN?
> Each module in the dt pixel pipe has different input data from raw data without rgb information over an unlimited to a limited colorspace. So you need to train for each of those intermediate data even those you cant display on a common display ;)
> keep in mind: in darktable you never see the output od a module - just the output of the whole pixelpipe

DNN has short life time and incrementally build with user interaction - user clicks on pixels, marking them as in/out filter, until it is satisfied with approximation - click is cheap. So DNN is rebuild after some module updates down the stack.

DNN is simple and training sets is small (hundred click max), that why it is can be incremental.

DNN input is a matrix of adjacent pixels 3x3 (in main use case) => 9 pixels for every pixel all channels (3 channels). Particular color space doesn't matter, because the encoded color is just a number vector. 

Color spaces are isomorphic, aren't they? The richest colorspace could be used to mitigate difference between color space used for training and used for running if it is presented. 

Code for DNN is as complex as a student's lab of AI course.
100 lines of tensorflow API

**Comment by yaitskov** [comment]
> Probably related: #12295.

wow - just 5 days ahead of me - must be some psychic  connection ;)
